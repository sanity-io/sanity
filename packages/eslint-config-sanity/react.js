// Use this for react projects
module.exports = {
  env: {
    node: true,
    browser: true,
  },
  plugins: ['react'],
  parserOptions: {
    ecmaFeatures: {
      jsx: true,
    },
  },
  rules: {
    'react/boolean-prop-naming': 'off', // todo: enable this with list of native html boolean attributes allowed (and is/has/can/will/*-prefixes)
    'react/button-has-type': 'error',
    'react/display-name': ['error', {ignoreTranspilerName: false}],
    'react/default-props-match-prop-types': 'error',
    'react/destructuring-assignment': 'off',
    'react/forbid-elements': 'off',
    'react/forbid-foreign-prop-types': 'error',
    'react/forbid-component-props': 'off',
    'react/forbid-dom-props': 'off',
    'react/forbid-prop-types': 'warn',
    'react/jsx-no-comment-textnodes': 'error',
    'react/jsx-sort-props': 'off',
    'react/jsx-first-prop-new-line': ['error', 'multiline'],
    'react/jsx-no-target-blank': 'error',
    'react/jsx-filename-extension': ['error', {extensions: ['.js']}],
    'react/jsx-boolean-value': 'error',
    'react/jsx-child-element-spacing': 'off', // missing docs atm
    'react/jsx-closing-bracket-location': 1,
    'react/jsx-closing-tag-location': 'error',
    'react/jsx-curly-brace-presence': 'off',
    'react/jsx-curly-spacing': ['error', 'never'],
    'react/jsx-equals-spacing': ['error', 'never'],
    'react/jsx-handler-names': 'error',
    'react/jsx-indent': ['error', 2],
    'react/jsx-indent-props': ['error', 2],
    'react/jsx-key': 'error',
    'react/jsx-max-props-per-line': ['error', {maximum: 6}],
    'react/jsx-no-bind': 'warn',
    'react/jsx-no-duplicate-props': 'error',
    'react/jsx-no-literals': 'off',
    'react/jsx-no-undef': 'error',
    'react/jsx-one-expression-per-line': 'off',
    'react/jsx-pascal-case': 'error',
    'react/jsx-sort-default-props': 'off',
    // "react/jsx-space-before-closing": ["error", "always"], //@@@ removed in favor of react/jsx-tag-spacing
    'react/jsx-uses-react': 'error',
    'react/jsx-uses-vars': 'error',
    'react/jsx-tag-spacing': [
      'error',
      {
        closingSlash: 'never',
        beforeSelfClosing: 'always',
        afterOpening: 'never',
      },
    ],
    'react/jsx-wrap-multilines': 'error',
    'react/no-access-state-in-setstate': 'error',
    'react/no-array-index-key': 'warn',
    'react/no-children-prop': 'error',
    // "react/no-comment-textnodes": "error",  //@@@ deprecated in favor of react/jsx-no-comment-textnodes
    'react/no-danger': 'warn',
    'react/no-danger-with-children': 'error',
    'react/no-deprecated': 'error',
    'react/no-did-mount-set-state': 'warn',
    'react/no-did-update-set-state': 'error',
    'react/no-direct-mutation-state': 'error',
    'react/no-find-dom-node': 'warn',
    'react/no-is-mounted': 'error',
    'react/no-multi-comp': 'off',
    'react/no-redundant-should-component-update': 'error',
    'react/no-render-return-value': 'error',
    'react/no-set-state': 'off',
    'react/no-string-refs': 'warn',
    'react/no-this-in-sfc': 'error',
    'react/no-typos': 'error',
    'react/no-unescaped-entities': ['error', {forbid: ['>', '}']}],
    'react/no-unknown-property': 'error',
    'react/no-unused-prop-types': 'error',
    'react/no-unused-state': 'error',
    'react/no-will-update-set-state': 'error',
    'react/prefer-es6-class': 'off',
    'react/prop-types': 'error',
    'react/react-in-jsx-scope': 'error',
    // "react/require-extension": "...", //@@@ deprecated in favor of the eslint-plugin-import extensions rule
    'react/require-optimization': 'off',
    'react/require-render-return': 'error',
    'react/require-default-props': 'warn',
    'react/prefer-stateless-function': ['warn', {ignorePureComponents: true}],
    'react/sort-comp': 'off',
    'react/sort-prop-types': 'off',
    'react/self-closing-comp': 'error',
    'react/style-prop-object': 'error',
    'react/void-dom-elements-no-children': 'error',

    // "react/wrap-multilines": "error" //@@@ deprecated in favor of react/jsx-wrap-multilines
  },
}
