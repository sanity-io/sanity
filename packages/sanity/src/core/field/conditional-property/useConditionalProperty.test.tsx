import {renderHook} from '@testing-library/react'
import {createTestProvider} from '../../../../test/testUtils/TestProvider'
import {
  ConditionalPropertyProps,
  unstable_useConditionalProperty as useConditionalProperty,
} from './useConditionalProperty'

const dummyDocument = {
  _createdAt: '2021-11-04T15:41:48Z',
  _id: 'drafts.10053a07-8647-4ebd-9d1d-33a512d30d3a',
  _rev: '5hb8s6-k75-ip4-4bq-5ztbf3fbx',
  _type: 'conditionalFieldsTest',
  _updatedAt: '2021-11-05T12:34:29Z',
  title: 'Hello world',
  isPublished: true,
  venue: {
    location: {lat: 37.813563, lng: -122.268812},
    address: {
      street: '2421 Telegraph Ave #102',
      city: 'Oakland',
      zip: 94612,
      state: 'California',
    },
  },
}

const DEFAULT_PROPS: Omit<ConditionalPropertyProps, 'checkProperty'> = {
  checkPropertyKey: 'testKey',
  document: dummyDocument,
  value: undefined,
  parent: {
    parentTest: 'hello',
    siblingProp: true,
  },
}

afterEach(() => {
  jest.resetAllMocks()
})

describe('Conditional property resolver', () => {
  /* eslint-disable max-nested-callbacks */
  it('calls callback function', async () => {
    const TestWrapper = await createTestProvider()
    const callbackFn = jest.fn(() => true)

    renderHook(
      () =>
        useConditionalProperty({
          checkProperty: callbackFn,
          ...DEFAULT_PROPS,
        }),
      {wrapper: TestWrapper},
    )

    expect(callbackFn).toBeCalled()

    expect(callbackFn.mock.calls).toMatchSnapshot()
  })

  it('resolves callback to true', async () => {
    const TestWrapper = await createTestProvider()
    const {result} = renderHook(
      () =>
        useConditionalProperty({
          checkProperty: jest.fn(() => true),
          ...DEFAULT_PROPS,
        }),
      {wrapper: TestWrapper},
    )
    expect(result.current).toBeTruthy()
  })

  it('returns false with callback that returns false', async () => {
    const TestWrapper = await createTestProvider()
    const {result} = renderHook(
      () =>
        useConditionalProperty({
          checkProperty: jest.fn(() => false),
          ...DEFAULT_PROPS,
        }),
      {wrapper: TestWrapper},
    )
    expect(result.current).toBe(false)
  })

  it('returns false if document title does not match', async () => {
    const TestWrapper = await createTestProvider()
    const {result} = renderHook(
      () =>
        useConditionalProperty({
          checkProperty: jest.fn(({document}) => document?.title !== 'Hello world'),
          ...DEFAULT_PROPS,
        }),
      {wrapper: TestWrapper},
    )
    expect(result.current).toBeFalsy()
  })

  it('returns true if document is published', async () => {
    const TestWrapper = await createTestProvider()
    const {result} = renderHook(
      () =>
        useConditionalProperty({
          checkProperty: jest.fn(({document}) => Boolean(document?.isPublished)),
          ...DEFAULT_PROPS,
        }),
      {wrapper: TestWrapper},
    )
    expect(result.current).toBeTruthy()
  })

  it('returns undefined because callback returns undefined', async () => {
    const consoleSpy = jest.spyOn(console, 'warn').mockImplementation(() => undefined)
    const TestWrapper = await createTestProvider()
    const {result} = renderHook(
      () =>
        useConditionalProperty({
          checkProperty: jest.fn(() => undefined) as any,
          ...DEFAULT_PROPS,
        }),
      {wrapper: TestWrapper},
    )
    expect(result.current).toBe(undefined)
    expect(consoleSpy).toHaveBeenCalledWith(
      'The `testKey` option is or returned `undefined`. `testKey` should return a boolean.',
    )
  })

  it('returns true because value matches', async () => {
    const TestWrapper = await createTestProvider()
    const {result} = renderHook(
      () =>
        useConditionalProperty({
          checkProperty: jest.fn(({value}) => value === 'test value'),
          ...DEFAULT_PROPS,
          value: 'test value',
        }),
      {wrapper: TestWrapper},
    )
    expect(result.current).toBeTruthy()
  })

  it('returns false because value does not match', async () => {
    const TestWrapper = await createTestProvider()
    const {result} = renderHook(
      () =>
        useConditionalProperty({
          checkProperty: jest.fn(({value}) => value === 'test'),
          ...DEFAULT_PROPS,
          value: 'test value',
        }),
      {wrapper: TestWrapper},
    )
    expect(result.current).toBeFalsy()
  })

  it('returns true when the current user does not have role "developer"', async () => {
    const TestWrapper = await createTestProvider()
    const {result} = renderHook(
      () =>
        useConditionalProperty({
          checkProperty: jest.fn(
            ({currentUser}) => !currentUser?.roles.some((role) => role.name === 'developer'),
          ),
          ...DEFAULT_PROPS,
          value: 'test value',
        }),
      {wrapper: TestWrapper},
    )
    expect(result.current).toBeTruthy()
  })

  it('returns true when the current user has role "administrator"', async () => {
    const TestWrapper = await createTestProvider()
    const {result} = renderHook(
      () =>
        useConditionalProperty({
          checkProperty: jest.fn(({currentUser}) =>
            Boolean(currentUser?.roles.some((role) => role.name === 'administrator')),
          ),
          ...DEFAULT_PROPS,
          value: 'test value',
        }),
      {wrapper: TestWrapper},
    )
    expect(result.current).toBeTruthy()
  })

  it('returns true when sibling field is not empty', async () => {
    const TestWrapper = await createTestProvider()
    const {result} = renderHook(
      () =>
        useConditionalProperty({
          checkPropertyKey: 'hidden',
          document: dummyDocument,
          value: dummyDocument.venue.address,
          parent: dummyDocument.venue,
          checkProperty: jest.fn(({parent}) => Boolean(parent.location)),
        }),
      {wrapper: TestWrapper},
    )
    expect(result.current).toBeTruthy()
  })
  /* eslint-enable max-nested-callbacks */
})
