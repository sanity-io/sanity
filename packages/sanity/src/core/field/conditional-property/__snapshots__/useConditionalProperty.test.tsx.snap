// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Conditional property resolver calls callback function 1`] = `
Array [
  Array [
    Object {
      "currentUser": Object {
        "email": "doug@sanity.io",
        "id": "doug",
        "name": "Doug",
        "role": "admin",
        "roles": Array [
          Object {
            "name": "administrator",
            "title": "Administrator",
          },
        ],
      },
      "document": Object {
        "_createdAt": "2021-11-04T15:41:48Z",
        "_id": "drafts.10053a07-8647-4ebd-9d1d-33a512d30d3a",
        "_rev": "5hb8s6-k75-ip4-4bq-5ztbf3fbx",
        "_type": "conditionalFieldsTest",
        "_updatedAt": "2021-11-05T12:34:29Z",
        "isPublished": true,
        "title": "Hello world",
        "venue": Object {
          "address": Object {
            "city": "Oakland",
            "state": "California",
            "street": "2421 Telegraph Ave #102",
            "zip": 94612,
          },
          "location": Object {
            "lat": 37.813563,
            "lng": -122.268812,
          },
        },
      },
      "parent": Object {
        "parentTest": "hello",
        "siblingProp": true,
      },
      "value": undefined,
    },
  ],
]
`;
